Increase bo cookie lifetime from 15 to 60 minutes

jf-viguier opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Merchants reports annoying deconnections all day because of too short & hard coded cookie lifetime in bo.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | Does it break backward compatibility? no
| Deprecations? | Does it deprecate an existing feature? no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/16877| How to test?  | Login and take a coffee break

This trick is mentioned here by @khouloudbelguith and in others issues. Why not include it in the core ?
https://github.com/PrestaShop/PrestaShop/issues/16877#issuecomment-567846712

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/16880)
<!-- Reviewable:end -->

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes

matthieu-rolland reviewed (commented):
Hi, thank you for your contribution 👍 

I don't see anything wrong with your contribution, but I wonder if this is a patch or a real fix of the described issue? 

Could you explain how this modification will keep the bug from happening again?

Shouldn't it take the value specified in the BO instead of being hardcoded ?

jf-viguier committed:
uses PS_COOKIE_LIFETIME_BO if set

jf-viguier committed:
uses PS_COOKIE_LIFETIME_BO if set

jf-viguier committed:
uses PS_COOKIE_LIFETIME_BO if set

jf-viguier committed:
uses PS_COOKIE_LIFETIME_BO if set

jf-viguier commented:
@matthieu-rolland I've made a new commit to uses PS_COOKIE_LIFETIME_BO var.
This fix allows to not been disconnected too often.

PierreRambaud commented on classes/controller/AdminController.php:
```diff
@@ -2831,7 +2831,7 @@ public function init()
         }
         if (isset(Context::getContext()->cookie->last_activity)) {
-            if ($this->context->cookie->last_activity + 900 < time()) {
+            if ($this->context->cookie->last_activity + max((int) Configuration::get('PS_COOKIE_LIFETIME_BO') * 3600, 3600) < time()) {
```
```suggestion
            $cookieLifetime = (int) Configuration::get('PS_COOKIE_LIFETIME_BO');
            if ($this->context->cookie->last_activity + (!empty($cookieLifetime) ? $cookieLifetime : static::DEFAULT_COOKIE_LIFETIME) < time()) {
```

and add the const `DEFAULT_COOKIE_LIFETIME` :)

jf-viguier replied:
where do I add DEFAULT_COOKIE_LIFETIME ? 3600 s is ok for u ?

PierreRambaud replied:
Could be in AdminController, or in Configuration, I've no idea what's the best :sweat_smile:

jf-viguier replied:
Most constants are in config\defines.inc.php

jf-viguier replied:
It's done @PierreRambaud with a new constant

PierreRambaud replied:
I'm not in favor of adding global const in defines.inc.php, it's pretty old school 😅 
Futhermore, why using max? What if I want to configure the cookie lifetime with 10 minutes, I need to edit defined.inc.php and data in the database? it's weird 🤔 
I think you must remove the max and use a class constant.

jf-viguier committed:
Add a constant for default cookie lifetime

jf-viguier committed:
Add a constant for default cookie lifetime

jf-viguier committed:
Add a constant for default cookie lifetime

jf-viguier committed:
Add a constant for default cookie lifetime

jf-viguier commented:
@PierreRambaud @matthieu-rolland We are on a wrong way. The back office PS_COOKIE_LIFETIME_BO variable should only used when the user choose "keep me logged in".
Let's check this issue:
https://github.com/PrestaShop/PrestaShop/issues/12043

So, I suggest to only change 15 to 60 minutes as it's my first suggestion. What do you think ?
@eternoendless has say it's ok : https://github.com/PrestaShop/PrestaShop/issues/12043#issuecomment-454841747

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes with a class constant.
Remove old school constant in defines.inc.php
The back office PS_COOKIE_LIFETIME_BO variable should only used when the user choose "keep me logged in".

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes with a class constant.
Remove old school constant in defines.inc.php
The back office PS_COOKIE_LIFETIME_BO variable should only used when the user choose "keep me logged in".

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes with a class constant.
Remove old school constant in defines.inc.php
The back office PS_COOKIE_LIFETIME_BO variable should only used when the user choose "keep me logged in".

jf-viguier committed:
Increase bo cookie lifetime from 15 to 60 minutes with a class constant.
Remove old school constant in defines.inc.php
The back office PS_COOKIE_LIFETIME_BO variable should only used when the user choose "keep me logged in".

jf-viguier committed:
PHP-CS-Fixer

jf-viguier committed:
PHP-CS-Fixer

jf-viguier committed:
PHP-CS-Fixer

jf-viguier committed:
PHP-CS-Fixer

jf-viguier commented:
I'e changed the delay from 15 to 60 minutes in a class constant.
I'm not sure of my rebase, can you help on this ?

PierreRambaud commented:
Look like there is a problem with the commits history :thinking:

jf-viguier commented:
@PierreRambaud how can I fix my commits history issue ?

PierreRambaud commented:
@jf-viguier By doing a git rebase from develop branch.

jf-viguier commented:
@PierreRambaud I've just done the rebase, is it ok now ?

PierreRambaud commented:
@jf-viguier Unfortunately no :sweat_smile: 
![image](https://user-images.githubusercontent.com/1462701/72419670-934da380-377d-11ea-8a6e-b751c78af80a.png)
Not sure you included so much code in your pull request.
Don't worry I'll do it for you :wink:

PierreRambaud commented:
Should be ok now! Could you tell me if everything's still working as expected?

jf-viguier commented:
The line has to be : if ($this->context->cookie->last_activity + self::AUTH_COOKIE_LIFETIME < time()) {
I've done a new commit that generated a merge commit, I don't understand why is there so much changes.
My modification is just 2 lines of code.
Tell me if it's ok.

jf-viguier commented:
I don't know why bo_theme has been deleted, it's not my modification

PierreRambaud commented:
because you didn't retrieve modification from branches properly :thinking: 
The line changed because I was thinking, getting the value from the BO could be better than having a non configurable data :thinking:

jf-viguier commented:
OK. Whqt do I need to fix this git issue ? Do I create a new branch and restart ?

Progi1984 commented:
@jf-viguier Hi, a solution would be a rebase. But if your PR is small and you're not skilled with rebase, you can create directly a new PR and close this one.

jf-viguier commented:
@Progi1984 I've already done a rebase, but I can't figure out why it's not ok

jf-viguier commented:
I've just done a rebase but it's a mess, I hat conflits and weirds modifications. I'll try a new pr

jf-viguier commented:
new pr https://github.com/PrestaShop/PrestaShop/pull/17497
